components/server/src/prebuilds/prebuild-manager.ts

```diff
@@ -76,6 +87,155 @@ export class PrebuildManager {
+            try {
+                const toDispose = new DisposableCollection();
+                for (const configurationId of configurationIds) {
+                    toDispose.push(
+                        this.subscriber.listenForPrebuildUpdates(configurationId, (_ctx, prebuild) => {
+                            sink.push(prebuild);
+                        }),
+                    );
+                }
+                return () => {
+                    toDispose.dispose();
+                };
+            } catch (e) {
+                if (e instanceof Error) {
+                    sink.fail(e);
+                    return;
+                } else {
+                    sink.fail(new Error(String(e) || "unknown"));
+                }
+            }
+        }, opts);
+    }
+
+    async triggerPrebuild(ctx: TraceContext, user: User, projectId: string, branchName: string | null) {
+        await this.auth.checkPermissionOnProject(user.id, "write_prebuild", projectId);
+
+        const project = await this.projectService.getProject(user.id, projectId);
+
+        const branchDetails = !!branchName
+            ? await this.projectService.getBranchDetails(user, project, branchName)
+            : (await this.projectService.getBranchDetails(user, project)).filter((b) => b.isDefault);
+        if (branchDetails.length !== 1) {
+            log.debug({ userId: user.id }, "Cannot find branch details.", { project, branchName });
+            throw new ApplicationError(
+                ErrorCodes.NOT_FOUND,
+                `Could not find ${!branchName ? "a default branch" : `branch '${branchName}'`} in repository ${
+                    project.cloneUrl
+                }`,
+            );
+        }
+        const contextURL = branchDetails[0].url;
+
+        const context = (await this.contextParser.handle(ctx, user, contextURL)) as CommitContext;
+
+        // HACK: treat manual triggered prebuild as a reset for the inactivity state
+        await this.projectService.markActive(user.id, project.id);
+
+        const prebuild = await this.startPrebuild(ctx, {
+            context,
+            user,
+            project,
+            forcePrebuild: true,
+        });
+
+        this.analytics.track({
+            userId: user.id,
+            event: "prebuild_triggered",
+            properties: {
+                context_url: contextURL,
+                clone_url: project.cloneUrl,
+                commit: context.revision,
+                branch: branchDetails[0].name,
+                project_id: project.id,
+            },
+        });
+
+        return prebuild;
+    }
+
+    async cancelPrebuild(
+        ctx: TraceContext,
+        userId: string,
+        prebuildId: string,
+        oldPermissionCheck?: (projectId: string) => Promise<void>, // @deprecated
+    ): Promise<void> {
+        const prebuild = await this.workspaceDB.trace(ctx).findPrebuildByID(prebuildId);
```

Comment:
This one is not moved as is, the thing is that JSON-RCP server has complicated internalStopWorkspace method which should work for workspace and prebuild, but at the end it calls stopWorkspace for Prebuild.

I decided not to change the server because of it to make sure that it works like before, and this one is used only with gRPC